src/ast/ast.h

```diff
@@ -62,6 +62,17 @@ enum class Operator {
   BNOT,
 };
 
+// There are 2 kinds of attach point expansion:
+// - full expansion  - separate LLVM function is generated for each match
+// - multi expansion - one LLVM function and BPF program is generated for all
+//                     matches, the list of expanded functions is attached to
+//                     the BPF program using the k(u)probe.multi mechanism
+enum class ExpansionType {
+  NONE,
+  FULL,
+  MULTI,
+};
```

Comment:
Could we detach the `k(u)probe.multi` feature from the `ExpansionType`?
What I'd love to be able to do is tell `bpftrace` which addresses to trap and it decides by itself if whether to use `k(u)probe.multi`, if available, or set-up the probe one-by-one for each address, if there are no multi probe support.

Reply:
This is an excellent point. If we do AOT, we may compile the script on a machine with kprobe_multi support but then run it on a machine which doesn't have it. In such a case, the execution would probably not work.

This is quite tricky to implement, though. Since we're about to use libbpf to load the programs, for full expansion, we need to have one copy of the function for each match. I'll work on a libbpf extension to support copies via aliases (i.e. just entries in the symbol table pointing to the same code) but for now, we need full copies.

The only reasonable option that comes to mind is to replace the `MULTI` expansion by `ALIAS` (or better name if we can find one). For that expansion type, we would always emit just one copy of the program. Then, the runtime would detect if `k(u)probe_multi` is available and, if not, create the copies directly in the ELF. In future, we could replace that by just creating aliasing symbol table entries. I'll just need to check how complicated it is to do symbol copying in ELF.

In the meantime, if someone has a better idea, let me know.